File: cli/src/session.rs
//! Config-protocol framing over a [`Probe`] transport.
//!
//! Each method issues one or more request/response packets, echoing the
//! command byte and checking the status byte (see `probe_config::protocol`).

use anyhow::{bail, Context, Result};
use probe_config::protocol::*;
use probe_config::Topology;

use crate::transport::Probe;

/// Max topology bytes per SET chunk (64-byte packet minus cmd + offset).
/// Mirrors the firmware's GET chunk size so a full topology fits either way.
const SET_CHUNK: usize = 61;

/// A config session bound to one opened probe interface.
pub struct Session {
    probe: Probe,
}

impl Session {
    /// Wrap an opened probe.
    pub fn new(probe: Probe) -> Self {
        Self { probe }
    }

    /// Read [`FirmwareInfo`] (`CMD_INFO`).
    pub fn info(&self) -> Result<FirmwareInfo> {
        let resp = self.probe.transceive(&[CMD_INFO])?;
        check(&resp, CMD_INFO)?;
        postcard::from_bytes(&resp[2..]).context("decode FirmwareInfo")
    }

    /// Read the active topology (chunked `CMD_GET_TOPOLOGY`).
    pub fn get_topology(&self) -> Result<Topology> {
        let mut buf: Vec<u8> = Vec::new();
        loop {
            let resp = self
                .probe
                .transceive(&[CMD_GET_TOPOLOGY, buf.len() as u8])?;
            check(&resp, CMD_GET_TOPOLOGY)?;
            if resp.len() < 3 {
                bail!("short GET_TOPOLOGY response ({} bytes)", resp.len());
            }
            let total_len = resp[2] as usize;
            let chunk = &resp[3..];
            if buf.len() >= total_len {
                break;
            }
            if chunk.is_empty() {
                bail!("probe returned an empty chunk before the topology was complete");
            }
            buf.extend_from_slice(chunk);
            if buf.len() >= total_len {
                buf.truncate(total_len);
                break;
            }
        }
        postcard::from_bytes(&buf).context("decode Topology")
    }

    /// Stage a topology (chunked `CMD_SET_TOPOLOGY`) and `CMD_COMMIT` it.
    /// The firmware re-validates before writing flash.
    pub fn set_topology(&self, topo: &Topology) -> Result<()> {
        let encoded = postcard::to_stdvec(topo).context("encode Topology")?;
        if encoded.len() > TOPOLOGY_BUF_LEN {
            bail!(
                "encoded topology is {} bytes, firmware staging buffer holds {}",
                encoded.len(),
                TOPOLOGY_BUF_LEN
            );
        }
        // Offset 0 resets the staging buffer, so send it first even when empty.
        let mut offset = 0;
        loop {
            let end = (offset + SET_CHUNK).min(encoded.len());
            let mut req = Vec::with_capacity(2 + (end - offset));
            req.push(CMD_SET_TOPOLOGY);
            req.push(offset as u8);
            req.extend_from_slice(&encoded[offset..end]);
            let resp = self.probe.transceive(&req)?;
            check(&resp, CMD_SET_TOPOLOGY)?;
            offset = end;
            if offset >= encoded.len() {
                break;
            }
        }
        let resp = self.probe.transceive(&[CMD_COMMIT, encoded.len() as u8])?;
        check(&resp, CMD_COMMIT)?;
        Ok(())
    }

    /// Reboot the probe (`CMD_REBOOT`); it re-enumerates afterward.
    pub fn reboot(&self) -> Result<()> {
        let resp = self.probe.transceive(&[CMD_REBOOT])?;
        check(&resp, CMD_REBOOT)?;
        Ok(())
    }
}

/// Human-readable meaning of a protocol status byte.
fn status_str(status: u8) -> &'static str {
    match status {
        STATUS_OK => "ok",
        STATUS_ERR_DECODE => "payload failed to decode",
        STATUS_ERR_INVALID => "topology failed validation",
        STATUS_ERR_FLASH => "flash write failed",
        STATUS_ERR_BAD_REQUEST => "unknown command or malformed arguments",
        _ => "unknown status",
    }
}

/// Verify a response echoes `cmd` and carries `STATUS_OK`.
fn check(resp: &[u8], cmd: u8) -> Result<()> {
    if resp.len() < 2 {
        bail!("short response ({} bytes)", resp.len());
    }
    if resp[0] != cmd {
        bail!(
            "response command 0x{:02X} does not echo request 0x{:02X}",
            resp[0],
            cmd
        );
    }
    if resp[1] != STATUS_OK {
        bail!(
            "probe reported error: {} (status 0x{:02X})",
            status_str(resp[1]),
            resp[1]
        );
    }
    Ok(())
}

File: webui/site/js/session.js
// Config-protocol framing over a transport, mirroring cli/src/session.rs.
//
// Command bytes, buffer sizes, and postcard encode/decode all come from the
// wasm module (the same probe-config crate the firmware runs), so this file
// only sequences packets.

const SET_CHUNK = 61; // 64-byte packet minus cmd + offset, mirrors the CLI

export class Session {
  /// `transport` from transport.js, `wasm` the initialized probe-config-wasm
  /// module.
  constructor(transport, wasm) {
    this.transport = transport;
    this.wasm = wasm;
    this.k = JSON.parse(wasm.constants());
  }

  statusStr(status) {
    const s = this.k.status;
    switch (status) {
      case s.ok: return "ok";
      case s.err_decode: return "payload failed to decode";
      case s.err_invalid: return "failed validation";
      case s.err_flash: return "flash write failed";
      case s.err_bad_request: return "unknown command or malformed arguments";
      default: return "unknown status";
    }
  }

  check(resp, cmd) {
    if (resp.length < 2) throw new Error(`short response (${resp.length} bytes)`);
    if (resp[0] !== cmd) {
      throw new Error(
        `response command 0x${resp[0].toString(16)} does not echo request 0x${cmd.toString(16)}`);
    }
    if (resp[1] !== this.k.status.ok) {
      throw new Error(
        `probe reported error: ${this.statusStr(resp[1])} (status 0x${resp[1].toString(16)})`);
    }
  }

  /// FirmwareInfo as a parsed object.
  async info() {
    const resp = await this.transport.transceive(new Uint8Array([this.k.cmd.info]));
    this.check(resp, this.k.cmd.info);
    return JSON.parse(this.wasm.decode_info(resp.subarray(2)));
  }

  /// Active topology as a JSON object {probes: [...], uarts: [...]}.
  async getTopology() {
    const cmd = this.k.cmd.get_topology;
    let buf = new Uint8Array(0);
    for (;;) {
      const resp = await this.transport.transceive(new Uint8Array([cmd, buf.length]));
      this.check(resp, cmd);
      if (resp.length < 3) throw new Error(`short GET_TOPOLOGY response (${resp.length} bytes)`);
      const totalLen = resp[2];
      const chunk = resp.subarray(3);
      if (buf.length >= totalLen) break;
      if (chunk.length === 0) {
        throw new Error("probe returned an empty chunk before the topology was complete");
      }
      const next = new Uint8Array(buf.length + chunk.length);
      next.set(buf); next.set(chunk, buf.length);
      buf = next;
      if (buf.length >= totalLen) { buf = buf.subarray(0, totalLen); break; }
    }
    return JSON.parse(this.wasm.decode_topology(buf));
  }

  /// Stage + commit a topology (JSON object). The firmware re-validates.
  async setTopology(topo) {
    const encoded = this.wasm.encode_topology(JSON.stringify(topo));
    const cmd = this.k.cmd.set_topology;
    // Offset 0 resets the staging buffer, so send it first even when empty.
    let offset = 0;
    for (;;) {
      const end = Math.min(offset + SET_CHUNK, encoded.length);
      const req = new Uint8Array(2 + (end - offset));
      req[0] = cmd; req[1] = offset;
      req.set(encoded.subarray(offset, end), 2);
      const resp = await this.transport.transceive(req);
      this.check(resp, cmd);
      offset = end;
      if (offset >= encoded.length) break;
    }
    const resp = await this.transport.transceive(
      new Uint8Array([this.k.cmd.commit, encoded.length]));
    this.check(resp, this.k.cmd.commit);
  }

  /// Current board profile as {available: "0-16,26-29", reserved: "16"}.
  async getProfile() {
    const cmd = this.k.cmd.get_profile;
    const resp = await this.transport.transceive(new Uint8Array([cmd]));
    this.check(resp, cmd);
    return JSON.parse(this.wasm.decode_profile(resp.subarray(2)));
  }

  /// Validate-and-store a board profile (JSON object).
  async setProfile(profile) {
    const encoded = this.wasm.encode_profile(JSON.stringify(profile));
    const cmd = this.k.cmd.set_profile;
    const req = new Uint8Array(1 + encoded.length);
    req[0] = cmd; req.set(encoded, 1);
    const resp = await this.transport.transceive(req);
    this.check(resp, cmd);
  }

  async reboot() {
    const resp = await this.transport.transceive(new Uint8Array([this.k.cmd.reboot]));
    this.check(resp, this.k.cmd.reboot);
  }

  async rebootBootsel() {
    const resp = await this.transport.transceive(new Uint8Array([this.k.cmd.bootsel]));
    this.check(resp, this.k.cmd.bootsel);
  }
}
